perses/app/experiments.py

```diff
@@ -0,0 +1,1301 @@
+            _simulation_flavor = edge_simulation_parameters[_phase][0]
+            _simulation_parameters = copy.deepcopy(edge_simulation_parameters[_phase][1])
+        else:
+            #check if we have a global_simulation_flavor
+            if type(self.global_simulation_flavor) == str:
+                assert self.simulation_parameters == None, f"if the global_simulation_flavor is set, then the self.simulation_parameters must be None"
+                #then we have a default
+                _simulation_flavor = self.global_simulation_flavor
+                _simulation_parameters = copy.deepcopy(self.simulation_arguments[_simulation_flavor])
+            else:
+                raise Exception(f"if the global simulation flavor is not set, then there are no appropriate defaults")
+
+        #modify the trajectory directory and trajectory_prefix
+        _simulation_parameters.update({'trajectory_directory': _simulation_parameters['trajectory_directory'].format(index0 = str(forward_index_map[0]), index1 = str(forward_index_map[1]))})
+        _simulation_parameters.update({'trajectory_prefix': _simulation_parameters['trajectory_prefix'].format(phase = phase)})
+
+        #now create a simulation object
+        simulation_object = Simulation(hybrid_factory = hybrid_factory,
+                                       sampler_type = _simulation_flavor,
+                                       sampler_arguments = _simulation_parameters)
+        return simulation_object
+
+
+    def _create_connectivity_matrix(self, graph_connectivity):
+        """
+        create a numpy 2d adjacency matrix for the digraph.
+
+        Arguments
+        ---------
+        graph_connectivity : str or np.array() (2D)
+            the key or explicit connectivity matrix with weights
+
+        Returns
+        -------
+        adjacency_matrix : np.array (2d)
+            adjacency matrix used to specify weights
+        """
+        if type(graph_connectivity) == str:
+            assert graph_connectivity in self.supported_connectivities, f"{graph_connectivity} is not supported.  Supported arguments are {self.supported_connectivities}."
+            return self.supported_connectivities[graph_connectivity](len(self.smiles_list))
+        elif type(graph_connectivity) == np.ndarray:
+            #we simply convert the weights into log weights
+            return np.log(graph_connectivity)
+
+
+    def _parse_ligand_input(self):
+        """
+        Parse the ligand input.
+        Creates the following attributes:
+        1. self.ligand_oemol_pos_top : list of tuple(oemol, position, topology)
+        2. self.ligand_ffxml : xml for ligands
+        3. self.smiles_list : list of smiles
+        4. self.ligand_md_topologies : list of mdtraj.Topology objects
+                                       corresponding to self.ligand_oemol_pos_top topologies.
+        """
+        self.ligand_oemol_pos_top = []
```

Comment:
This is a really weird data structure. Why have a tuple like this rather than storing these objects in more sensible data structures, like `self.molecule_oemols`, `self.molecule_positions`, and `self.molecule_topologies`? I think this just causes lots of confusion to group them in weird ways.

Reply:
for each molecule, we need to store a ligand oemol, positions, and openmm.Topology object.  after i created this, i realized we also needed to store a mdtraj.Topology, and forgot to make the it a 4-tuple, but i like the suggested lists more.  will change